superset/db_engine_specs/couchbasedb.py

```diff
@@ -0,0 +1,257 @@
+
+from datetime import datetime
+from typing import Any, Optional, TypedDict
+from urllib import parse
+
+from flask_babel import gettext as __
+from marshmallow import fields, Schema
+from sqlalchemy.engine.url import URL
+
+from superset.constants import TimeGrain
+from superset.databases.utils import make_url_safe
+from superset.db_engine_specs.base import (
+    BaseEngineSpec,
+    BasicParametersMixin,
+    BasicParametersType as BaseBasicParametersType,
+    BasicPropertiesType as BaseBasicPropertiesType,
+)
+from superset.errors import ErrorLevel, SupersetError, SupersetErrorType
+from superset.utils.network import is_hostname_valid, is_port_open
+
+
+class BasicParametersType(TypedDict, total=False):
+    username: Optional[str]
+    password: Optional[str]
+    host: str
+    database: str
+    port: Optional[int]
+    query: dict[str, Any]
+    encryption: bool
+
+
+class BasicPropertiesType(TypedDict):
+    parameters: BasicParametersType
+
+
+class CouchbaseParametersSchema(Schema):
+    username = fields.String(allow_none=True, metadata={"description": __("Username")})
+    password = fields.String(allow_none=True, metadata={"description": __("Password")})
+    host = fields.String(
+        required=True, metadata={"description": __("Hostname or IP address")}
+    )
+    database = fields.String(
+        allow_none=True, metadata={"description": __("Database name")}
+    )
+    port = fields.Integer(
+        allow_none=True, metadata={"description": __("Database port")}
+    )
+    encryption = fields.Boolean(
+        dump_default=False,
+        metadata={"description": __("Use an encrypted connection to the database")},
+    )
+    query = fields.Dict(
+        keys=fields.Str(),
+        values=fields.Raw(),
+        metadata={"description": __("Additional parameters")},
+    )
+
+
+class CouchbaseDbEngineSpec(BasicParametersMixin, BaseEngineSpec):
+    engine = "couchbasedb"
+    engine_name = "Couchbase"
+    default_driver = "couchbasedb"
+    allows_joins = False
+    allows_subqueries = False
+    sqlalchemy_uri_placeholder = (
+        "couchbasedb://user:password@host[:port]?truststorepath=value?ssl=value"
+    )
+    parameters_schema = CouchbaseParametersSchema()
+
+    _time_grain_expressions = {
+        None: "{col}",
+        TimeGrain.SECOND: "DATE_TRUNC_STR(TOSTRING({col}),'second')",
+        TimeGrain.MINUTE: "DATE_TRUNC_STR(TOSTRING({col}),'minute')",
+        TimeGrain.HOUR: "DATE_TRUNC_STR(TOSTRING({col}),'hour')",
+        TimeGrain.DAY: "DATE_TRUNC_STR(TOSTRING({col}),'day')",
+        TimeGrain.MONTH: "DATE_TRUNC_STR(TOSTRING({col}),'month')",
+        TimeGrain.YEAR: "DATE_TRUNC_STR(TOSTRING({col}),'year')",
+        TimeGrain.QUARTER: "DATE_TRUNC_STR(TOSTRING({col}),'quarter')",
+    }
+
+    @classmethod
+    def epoch_to_dttm(cls) -> str:
+        return "MILLIS_TO_STR({col} * 1000, '111')"
+
+    @classmethod
+    def epoch_ms_to_dttm(cls) -> str:
+        return "MILLIS_TO_STR({col}, '111')"
```

Comment:
I tried reading up on the docs, and I didn't fully understand how this format is valid:
- `MILLIS_TO_STR`: https://docs.couchbase.com/cloud/n1ql/n1ql-language-reference/datefun.html#fn-date-millis-to-str
- Date format: https://docs.couchbase.com/cloud/n1ql/n1ql-language-reference/datefun.html#date-string

To be clear, the idea of these methods is to provide an expression that can convert an epoch (seconds since 1970-01-01 00:00:00 in UTC) into a native temporal format. But if it works it works - if this is the case, maybe adding a comment as to why the format `'111'` works here would be helpful.

Reply:
Thank you so much for your comments,
If format is unspecified or incorrect it gives combined date and time in result , On querying 
SELECT MILLIS_TO_STR(1463284740000,'111');  it gives me "2016-05-15T03:59:00Z" in result . I had this '111' just for notation . But I have removed this in my new patch , because MILLIS_TO_STR(1463284740000) also gives me same results.